soroban-env-host/src/crypto/poseidon/poseidon2.rs

```diff
@@ -0,0 +1,267 @@
+    pub(crate) params: Poseidon2Params<F>,
+}
+
+impl<F: MeteredScalar> Poseidon2<F> {
+    pub fn new(params: Poseidon2Params<F>) -> Self {
+        Poseidon2 { params }
+    }
+
+    pub fn permutation(&self, host: &Host, input: &Vec<F>) -> Result<Vec<F>, HostError> {
+        let t = self.params.t;
+        if input.len() != t {
+            return Err(host.error(
+                INVALID_INPUT,
+                "poseidon2::permutation input length does not match `t`",
+                &[
+                    Val::from_u32(input.len() as u32).into(),
+                    Val::from_u32(t as u32).into(),
+                ],
+            ));
+        }
+
+        let mut current_state = input.metered_clone(host)?;
+
+        // Linear layer at beginning
+        self.matmul_external(host, &mut current_state)?;
+
+        for r in 0..self.params.rounds_f_beginning {
+            let rc = self.params.round_constants.get(r).ok_or_else(|| {
+                host.error(
+                    VEC_OOB,
+                    "poseidon2: round constant index out of bounds",
+                    &[],
+                )
+            })?;
+            current_state = utils::add_rc(host, &current_state, rc)?;
+            current_state = utils::sbox(host, &current_state, self.params.d)?;
+            self.matmul_external(host, &mut current_state)?;
+        }
+
+        let p_end = self.params.rounds_f_beginning + self.params.rounds_p;
+        for r in self.params.rounds_f_beginning..p_end {
+            let rc = self.params.round_constants.get(r).ok_or_else(|| {
+                host.error(
+                    VEC_OOB,
+                    "poseidon2: round constant index out of bounds",
+                    &[],
+                )
+            })?;
+            let rc_first = rc.first().ok_or_else(|| {
+                host.error(VEC_OOB, "poseidon2: round constant vector is empty", &[])
+            })?;
+            let first = current_state
+                .first_mut()
+                .ok_or_else(|| host.error(VEC_OOB, "poseidon2: current_state is empty", &[]))?;
+            first.metered_add_assign(rc_first, host)?;
+            *first = utils::sbox_p(host, first, self.params.d)?;
+            self.matmul_internal(host, &mut current_state, &self.params.mat_internal_diag_m_1)?;
+        }
+
+        for r in p_end..self.params.rounds {
+            let rc = self.params.round_constants.get(r).ok_or_else(|| {
+                host.error(
+                    VEC_OOB,
+                    "poseidon2: round constant index out of bounds",
+                    &[],
+                )
+            })?;
+            current_state = utils::add_rc(host, &current_state, rc)?;
+            current_state = utils::sbox(host, &current_state, self.params.d)?;
+            self.matmul_external(host, &mut current_state)?;
+        }
+        Ok(current_state)
+    }
+
+    fn matmul_m4(&self, host: &Host, input: &mut [F]) -> Result<(), HostError> {
```

Comment:
A comment on what this does would be useful. Also, is there some implementation/specification reference for this?